File: RB/ts/RiotBase/Variable/ObjectId.ts
import Variable from "./Core/Variable"

//
export default
class ObjectId extends Variable {
    constructor( val? ) {
        if      ( val instanceof Rp         ) super( val );
        else if ( val instanceof ObjectId   ) super( this.copy().rp );
        else if ( val instanceof Inode      ) super( new ObjectId_Inode( val ) );
        else if ( val == undefined          ) super( new ObjectId_Inode( new Inode ) );
        else throw new Error( "String from a " + val.constructor.name );
    }

    get() : Inode { // to  native type
         return this.toInode();
    }

    copy() : ObjectId {
        return Method.methods["copy"].call_1( this );
    }

    toInode() : Inode {
        return Method.methods["to_Inode"].call_1( this );
    }
}

import ObjectId_Inode from "./ObjectId/ObjectId_Inode";
import Inode          from "../System/Inode";
import Method         from "./Core/Method";
import Rp             from "./Core/Rp";

File: RB/cpp/RiotBase/Variable/ObjectId.cpp
#include "ObjectId/ObjectId_Inode.h"
#include "ObjectId/OtWrapperInode.h"
#include "ObjectId.h"

namespace RiotBase {

ObjectId::ObjectId( Variable::CreatorId, UsrId usr_id ) {
    if ( usr_id )
        rp = OtWrapperInode::New( Inode{}, usr_id );
    else
        rp = ObjectId_Inode::New( Inode{} );
}

ObjectId::ObjectId( const ObjectId &id ) : ObjectId( RO_CALL( copy, id.rp ) ) {
}

ObjectId::ObjectId( Inode inode ) : ObjectId( ObjectId_Inode::New( inode ) ) {
}

bool ObjectId::to_Inode( Inode &res ) const {
    return RO_CALL( to_Inode, rp, &res, 1 );
}

std::string ObjectId::category_name() {
    return "ObjectId";
}

} // namespace RiotBase
